File: backend/services/razorpay_service.py
"""Razorpay integration (real + dev-mode)."""
from __future__ import annotations
import os
import hmac
import hashlib
import secrets
import time
import logging

logger = logging.getLogger(__name__)


def is_dev_mode() -> bool:
    from services import settings_service
    return settings_service.get_bool("razorpay", "dev_mode", "RAZORPAY_DEV_MODE", default=False)


def _creds() -> tuple[str, str]:
    from services import settings_service
    return (
        settings_service.get_value("razorpay", "key_id", "RAZORPAY_KEY_ID"),
        settings_service.get_value("razorpay", "key_secret", "RAZORPAY_KEY_SECRET"),
    )


def _has_creds() -> bool:
    kid, ks = _creds()
    return bool(kid and ks)


def create_order(amount_paise: int, receipt: str, notes: dict | None = None) -> dict:
    if is_dev_mode() or not _has_creds():
        return {
            "id": f"order_dev_{secrets.token_hex(10)}",
            "amount": amount_paise, "currency": "INR",
            "receipt": receipt, "status": "created", "mock": True,
        }
    import razorpay
    kid, ks = _creds()
    client = razorpay.Client(auth=(kid, ks))
    order = client.order.create({"amount": amount_paise, "currency": "INR", "receipt": receipt, "notes": notes or {}})
    return order


def verify_signature(order_id: str, payment_id: str, signature: str) -> bool:
    if is_dev_mode() or not _has_creds():
        return True  # dev-mode: always succeed
    _, ks = _creds()
    secret = ks.encode()
    payload = f"{order_id}|{payment_id}".encode()
    expected = hmac.new(secret, payload, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)


def verify_webhook_signature(body: bytes, signature: str) -> bool:
    from services import settings_service
    wh = settings_service.get_value("razorpay", "webhook_secret", "RAZORPAY_WEBHOOK_SECRET")
    if is_dev_mode() or not wh:
        return True
    secret = wh.encode()
    expected = hmac.new(secret, body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)


def public_key_id() -> str:
    kid, _ = _creds()
    return kid or "rzp_test_dev_mock"

File: backend/src/services/razorpay.service.js
const crypto = require('crypto');
const config = require('../config');
const settingsService = require('./settings.service');
const logger = require('../utils/logger');

const isDevMode = () => {
  return settingsService.getBool('razorpay', 'dev_mode', 'RAZORPAY_DEV_MODE', false);
};

const getCreds = () => {
  return {
    keyId: settingsService.getValue('razorpay', 'key_id', 'RAZORPAY_KEY_ID'),
    keySecret: settingsService.getValue('razorpay', 'key_secret', 'RAZORPAY_KEY_SECRET'),
  };
};

const hasCreds = () => {
  const { keyId, keySecret } = getCreds();
  return !!(keyId && keySecret);
};

/**
 * Log the current Razorpay configuration status on startup.
 */
const logConfigStatus = () => {
  const dev = isDevMode();
  const creds = hasCreds();
  const { keyId } = getCreds();

  if (dev) {
    logger.warn('[Razorpay] DEV MODE is ON — all payments will be mocked. Set RAZORPAY_DEV_MODE=false for real payments.');
  } else if (!creds) {
    logger.error('[Razorpay] PRODUCTION MODE but credentials are MISSING. Set RAZORPAY_KEY_ID and RAZORPAY_KEY_SECRET in .env or admin settings.');
  } else {
    logger.info(`[Razorpay] Production mode active. Key ID: ${keyId.substring(0, 12)}...`);
  }
};

const createOrder = async (amountPaise, receipt, notes = {}) => {
  if (isDevMode()) {
    const devId = `order_dev_${crypto.randomBytes(10).toString('hex')}`;
    logger.info(`[Razorpay] DEV MODE: Created mock order ${devId} for ${amountPaise} paise`);
    return {
      id: devId,
      amount: amountPaise,
      currency: 'INR',
      receipt,
      status: 'created',
      mock: true,
    };
  }

  if (!hasCreds()) {
    logger.error('[Razorpay] Cannot create order: Razorpay credentials are not configured.');
    throw new Error('Payment gateway credentials are not configured. Please contact support.');
  }

  const { keyId, keySecret } = getCreds();

  try {
    const Razorpay = require('razorpay');
    const rzp = new Razorpay({ key_id: keyId, key_secret: keySecret });

    const order = await rzp.orders.create({
      amount: amountPaise,
      currency: 'INR',
      receipt,
      notes,
    });

    logger.info(`[Razorpay] Order created: ${order.id}, amount: ${amountPaise} paise, receipt: ${receipt}`);
    return order;
  } catch (err) {
    logger.error(`[Razorpay] Order creation FAILED: ${err.message}`, {
      statusCode: err.statusCode,
      error: err.error,
      amountPaise,
      receipt,
    });
    throw new Error(`Payment order creation failed: ${err.message}`);
  }
};

const verifySignature = (orderId, paymentId, signature) => {
  if (isDevMode()) {
    logger.warn('[Razorpay] DEV MODE: Skipping signature verification');
    return true;
  }

  if (!hasCreds()) {
    logger.error('[Razorpay] Cannot verify signature: credentials missing');
    return false;
  }

  const { keySecret } = getCreds();
  const shasum = crypto.createHmac('sha256', keySecret);
  shasum.update(`${orderId}|${paymentId}`);
  const digest = shasum.digest('hex');
  const isValid = digest === signature;

  if (!isValid) {
    logger.warn(`[Razorpay] Signature verification FAILED for order ${orderId}, payment ${paymentId}`);
  } else {
    logger.info(`[Razorpay] Signature verified for order ${orderId}, payment ${paymentId}`);
  }

  return isValid;
};

const verifyWebhookSignature = (bodyBytes, signature) => {
  const wh = settingsService.getValue('razorpay', 'webhook_secret', 'RAZORPAY_WEBHOOK_SECRET');
  if (isDevMode() || !wh) {
    return true;
  }
  const shasum = crypto.createHmac('sha256', wh);
  shasum.update(bodyBytes);
  const digest = shasum.digest('hex');
  return digest === signature;
};

const publicKeyId = () => {
  if (isDevMode()) {
    return 'rzp_test_dev_mock';
  }
  const { keyId } = getCreds();
  if (!keyId) {
    logger.error('[Razorpay] publicKeyId called but RAZORPAY_KEY_ID is empty');
  }
  return keyId || '';
};

module.exports = {
  isDevMode,
  createOrder,
  verifySignature,
  verifyWebhookSignature,
  publicKeyId,
  logConfigStatus,
};
